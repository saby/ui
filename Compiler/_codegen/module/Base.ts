/**
 */

import {
    IDependenciesController,
    createDependenciesController,
} from './Dependencies';
import { IExportsController, createExportsController } from './Exports';

/**
 * Интерфейс процессора, который выполняет компиляцию модуля в различный формат.
 */
export interface IModuleProcessor {
    /**
     * Установить имя компилируемого модуля.
     * @param name {string} Имя компилируемого модуля, которое содержит имя плагина.
     */
    setModuleName(name: string): void;

    /**
     * Установить флаг строго режима для компилируемого модуля.
     * @param value {boolean} Значение флага строго режима.
     */
    setStrictMode(value: boolean): void;

    /**
     * Добавить зависимость модуля. Если не указано имя зависимости, то зависимость считается неименованной.
     * @param ref {string} Путь зависимости.
     * @param name {string?} Имя зависимости.
     */
    addDependency(ref: string, name?: string): void;

    /**
     * Добавить блок код к телу компилируемого модуля.
     * @param block {string} Блок кода, который составляет тело модуля.
     */
    addCodeBlock(block: string): void;

    /**
     * Установить блок кода, который будет возвращен из модуля как экспортируемые данные.
     * @param block {string} Блок кода как экспортируемые данные.
     */
    setReturnableExport(block: string): void;

    /**
     * Выполнить компиляцию модуля.
     * @returns {string} JavaScript код модуля.
     */
    compile(): string;
}

/**
 * Абстрактный базовый класс процессора модулей.
 * Предоставляет общие данные и методы для выполнения компиляции.
 */
export abstract class Base implements IModuleProcessor {
    /**
     * Имя компилируемого модуля.
     * @protected
     */
    protected name: string;

    /**
     * Флаг включения строго режима для компилируемого модуля.
     * @protected
     */
    protected useStrict: boolean;

    /**
     * Массив кода, который составляет тело будущего модуля.
     * @protected
     */
    protected code: string[];

    /**
     * Инстанс контроллера зависимостей, необходимый для работы с именованными и
     * безымянными (анонимными) зависимостями модуля.
     * @protected
     */
    protected dependenciesController: IDependenciesController;

    /**
     * Инстанс контроллера экспортируемых данных.
     * @protected
     */
    protected exportsController: IExportsController;

    /**
     * Инициализировать инстанс базового класса.
     * @protected
     */
    protected constructor() {
        this.name = '';
        this.useStrict = false;
        this.code = [];
        this.dependenciesController = createDependenciesController();
        this.exportsController = createExportsController();
    }

    /**
     * Установить имя компилируемого модуля.
     * @param name {string} Имя компилируемого модуля, которое содержит имя плагина.
     */
    setModuleName(name: string): void {
        this.name = name;
    }

    /**
     * Установить флаг строго режима для компилируемого модуля.
     * @param value {boolean} Значение флага строго режима.
     */
    setStrictMode(value: boolean): void {
        this.useStrict = value;
    }

    /**
     * Добавить зависимость модуля. Если не указано имя зависимости, то зависимость считается неименованной.
     * @param ref {string} Путь зависимости.
     * @param name {string?} Имя зависимости.
     */
    addDependency(ref: string, name?: string): void {
        this.dependenciesController.addDependency(ref, name);
    }

    /**
     * Добавить блок код к телу компилируемого модуля.
     * @param block {string} Блок кода, который составляет тело модуля.
     */
    addCodeBlock(block: string): void {
        this.code.push(block);
    }

    /**
     * Установить блок кода, который будет возвращен из модуля как экспортируемые данные.
     * @param block {string} Блок кода как экспортируемые данные.
     */
    setReturnableExport(block: string): void {
        this.exportsController.setReturnableExport(block);
    }

    /**
     * Выполнить компиляцию модуля.
     * @abstract
     * @returns {string} JavaScript код модуля.
     */
    abstract compile(): string;
}
