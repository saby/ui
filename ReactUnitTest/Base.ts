import { Control } from 'UICore/Base';
import { IControlOptions } from 'UICommon/Base';

export interface ITestOptions extends IControlOptions {
    testName?: string;
    afterMountCallback?: Function;
    afterUpdateCallback?: Function;
}

export class TestBaseControl<
    IOptions extends ITestOptions = ITestOptions
> extends Control<IOptions> {
    testName: string = '';
    afterMountCallback: Function = null;
    afterUpdateCallback: Function = null;
    _beforeMount(options: ITestOptions): void {
        this.afterMountCallback = options.afterMountCallback;
        this.afterUpdateCallback = options.afterUpdateCallback;
        this.testName = options.testName;
    }
    _afterMount(): void {
        this.afterMountCallback?.(this);
    }
    _afterUpdate(): void {
        this.afterUpdateCallback?.(this);
    }
}
