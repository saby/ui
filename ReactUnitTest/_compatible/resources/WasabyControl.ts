import { Control, TemplateFunction } from 'UI/Base';

// @ts-ignore
import template = require('wml!ReactUnitTest/_compatible/resources/WasabyControl');

export default class WasabyControl extends Control {
    readonly _template: TemplateFunction = template;
}
