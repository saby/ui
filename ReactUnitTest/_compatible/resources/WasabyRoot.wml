<div>
    <Core.CompoundContainer component="ReactUnitTest/_compatible/resources/WS3Control" />
</div>