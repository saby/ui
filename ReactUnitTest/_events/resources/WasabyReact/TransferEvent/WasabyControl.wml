<ws:template name="hocReact">
    <ws:partial template="ReactUnitTest/_events/resources/WasabyReact/TransferEvent/ReactControl"
        on:click="_clickHandlerInTemplate()"
        on:mousedown="_mouseDownHandler()"
        on:myclick="_myClickHandler()"
        pushToQueue="{{pushToQueue}}"
    />
</ws:template>

<div>
    <ws:partial template="hocReact" on:click="_clickHandler()" />
    <button on:click="_applyEventQueue()" id="checkButton">check</button>
    <div>
        <ws:for data="value in _eventQueue">
            <div>
                {{value}}
            </div>
        </ws:for>
    </div>
</div>
