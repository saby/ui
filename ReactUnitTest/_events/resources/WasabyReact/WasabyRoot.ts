import { Control, TemplateFunction } from 'UI/Base';

// @ts-ignore
import template = require('wml!ReactUnitTest/_events/resources/WasabyReact/WasabyRoot');

export default class WasabyRoot extends Control {
    readonly _template: TemplateFunction = template;
}
