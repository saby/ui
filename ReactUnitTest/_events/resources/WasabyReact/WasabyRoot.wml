<div>
    <ReactUnitTest._events.resources.WasabyReact.ReactControl clickFn="{{ _options.clickFn }}" />
    <ReactUnitTest._events.resources.WasabyReact.ReactControlUsingHoc clickFn="{{ _options.clickFn }}" />
</div>
