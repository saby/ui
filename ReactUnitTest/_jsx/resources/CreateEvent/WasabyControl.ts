import { Control, TemplateFunction } from 'UI/Base';
/* eslint-disable no-console */
// @ts-ignore
import * as template from 'wml!ReactUnitTest/_jsx/resources/CreateEvent/WasabyControl';

export default class WasabyControl extends Control {
    protected _template: TemplateFunction = template;
}
