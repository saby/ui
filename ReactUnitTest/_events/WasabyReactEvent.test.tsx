/**
 * @jest-environment jsdom
 */
import { createSandbox } from 'sinon';
import { act } from 'react-dom/test-utils';

import { Control } from 'UICore/Base';
import WasabyRoot from './resources/WasabyReact/WasabyRoot';
import WasabyRootWithPropsOnClick from './resources/WasabyReact/WasabyRootWithPropsOnClick';
import WasabyControl from './resources/WasabyReact/TransferEvent/WasabyControl';

const creator = Control.createControl;
const destroyer = Control.destroyControl;

describe('WasabyReact Events', () => {
    let container;
    let sandbox;
    let clock;
    let consoleMock;
    let reactControlOnClickMock;
    let instance;

    beforeEach(() => {
        sandbox = createSandbox();
        clock = sandbox.useFakeTimers();
        consoleMock = jest.spyOn(console, 'error').mockImplementation();
        container = document.createElement('div');
        document.body.appendChild(container);
        reactControlOnClickMock = jest.fn();
    });

    afterEach(() => {
        clock.restore();
        sandbox.restore();
        consoleMock.mockRestore();
        destroyer(instance, container);
        container.remove();
        container = null;
        reactControlOnClickMock.mockClear();
    });

    function tick(duration: number): void {
        act(() => {
            clock.tick(duration);
        });
    }

    it('Проверяем события, если подписка добавлена через onClick на wasaby-контрол', () => {
        // region Setup
        act(() => {
            instance = creator(WasabyRoot, { clickFn: reactControlOnClickMock }, container);
        });
        tick(0);
        // endregion
        const button = document.getElementById('wasabyControl');

        act(() => {
            button.dispatchEvent(new window.MouseEvent('click', { bubbles: true }));
        });
        tick(0);

        expect(reactControlOnClickMock).toHaveBeenCalledTimes(1);
    });

    it('Проверяем события, если подписка добавлена через onClick на wasaby-контрол (HOC)', () => {
        // region Setup
        act(() => {
            instance = creator(WasabyRoot, { clickFn: reactControlOnClickMock }, container);
        });
        tick(0);
        // endregion
        const button = document.getElementById('wasabyControlHoc');
        act(() => {
            button.dispatchEvent(new window.MouseEvent('click', { bubbles: true }));
        });
        tick(0);

        expect(reactControlOnClickMock).toHaveBeenCalledTimes(1);
    });

    it('Проверяем события, если подписка добавлена через onClick на wasaby-контрол в корне которого partial с шаблоном', () => {
        // region Setup
        act(() => {
            instance = creator(WasabyRoot, { clickFn: reactControlOnClickMock }, container);
        });
        tick(0);
        // endregion
        const button = document.getElementById('wasabyControlTemplate');
        act(() => {
            button.dispatchEvent(new window.MouseEvent('click', { bubbles: true }));
        });
        tick(0);

        expect(reactControlOnClickMock).toHaveBeenCalledTimes(1);
    });

    it('Проверяем события, react-обработчик использует аргумент event', () => {
        // region Setup
        act(() => {
            instance = creator(WasabyRoot, { clickFn: reactControlOnClickMock }, container);
        });
        tick(0);
        // endregion
        const button = document.getElementById('wasabyControlWithArgs');
        act(() => {
            button.dispatchEvent(new window.MouseEvent('click', { bubbles: true }));
        });
        tick(0);

        expect(reactControlOnClickMock).toHaveBeenCalledTimes(1);
    });

    it('Проверяем события, react-обработчик использует аргумент event', () => {
        // region Setup
        act(() => {
            instance = creator(WasabyRoot, { clickFn: reactControlOnClickMock }, container);
        });
        tick(0);
        // endregion
        const button = document.getElementById('wasabyControlWithArgs2');
        act(() => {
            button.dispatchEvent(new window.MouseEvent('click', { bubbles: true }));
        });
        tick(0);

        expect(reactControlOnClickMock).toHaveBeenCalledTimes(1);
    });

    it('Проверяем события, если именование опции совпадает с именем реакт-события, но не является обработчиком', () => {
        // region Setup
        act(() => {
            instance = creator(WasabyRootWithPropsOnClick, {}, container);
        });
        tick(0);
        // endregion
        const button = document.getElementById('testButton');
        act(() => {
            button.dispatchEvent(new window.MouseEvent('click', { bubbles: true }));
        });
        tick(0);

        expect(consoleMock).not.toHaveBeenCalled();
        expect(container).toMatchSnapshot();
    });
});
