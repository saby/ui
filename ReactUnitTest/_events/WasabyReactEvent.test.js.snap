// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WasabyReact Events Проверяем события, если именование опции совпадает с именем реакт-события, но не является обработчиком 1`] = `
<div
  style="height: 100%; width: 100%; display: contents;"
>
  <div>
    <button
      id="testButton"
      tabindex="0"
    >
      true
    </button>
    <button
      id="testButton"
      tabindex="0"
    >
      123
    </button>
    <button
      id="testButton"
      tabindex="0"
    >
      test
    </button>
  </div>
</div>
`;
